text/0000-rebalancing-coherence.md

```diff
@@ -0,0 +1,293 @@
+patterns that the simple rules presented thus far can't handle.
+
+The first is that it is common to implement traits over boxed trait
+objects. For example, the `error` crate defines an impl:
+
+- `impl<E: Error> FromError<E> for Box<Error>`
+
+Here, `Error` is a local trait defined in `error`, but `FromError` is
+the trait from `libstd`. This impl would be illegal because
+`Box<Error>` is not considered local as `Box` is not local.
+
+The second is that it is common to use `FnMut` in blanket impls,
+similar to how the `Pattern` trait in `libstd` works. The `regex` crate
+in particular has the following impls:
+
+- `impl<'t> Replacer for &'t str`
+- `impl<F> Replacer for F where F: FnMut(&Captures) -> String`
+- these are in conflict because this requires that `&str: !FnMut`, and
+  neither `&str` nor `FnMut` are local to `regex`
+
+Given that overloading over closures is likely to be a common request,
+and that the `Fn` traits are well-known, core traits tied to the call
+operator, it seems reasonable to say that implementing a `Fn` trait is
+itself a breaking change. (This is not to suggest that there is
+something *fundamental* about the `Fn` traits that distinguish them
+from all other traits; just that if the goal is to have rules that
+users can easily remember, saying that implememting a core operator
+trait is a breaking change may be a reasonable rule, and it enables
+useful patterns to boot -- patterns that are baked into the libstd
+APIs.)
+
+To accommodate these cases (and future cases we will no doubt
+encounter), this RFC proposes an unstable attribute
+`#[fundamental]`. `#[fundamental]` can be applied to types and traits
+with the following meaning:
+
+- A `#[fundamental]` type `Foo` is one where implementing a blanket
+  impl over `Foo` is a breaking change. As described, `&` and `&mut` are
+  fundamental. This attribute would be applied to `Box`, making `Box`
+  behave the same as `&` and `&mut` with respect to coherence.
+- A `#[fundamental]` trait `Foo` is one where adding an impl of `Foo`
+  for an existing type is a breaking change. For now, the `Fn` traits
+  and `Sized` would be marked fundamental, though we may want to
+  extend this set to all operators or some other
+  more-easily-remembered set.
+
+The `#[fundamental]` attribute is intended to be a kind of "minimal
+commitment" that still permits the most important impl patterns we see
+in the wild. Because it is unstable, it can only be used within libstd
+for now. We are eventually committed to finding some way to
+accommodate the patterns above -- which could be as simple as
+stabilizing `#[fundamental]` (or, indeed, reverting this RFC
+altogether). It could also be a more general mechanism that lets users
+specify more precisely what kind of impls are reserved for future
+expansion and which are not.
+
+## Detailed Design
+
+### Proposed orphan rules
+
+Given an impl `impl<P1...Pn> Trait<T1...Tn> for T0`, either `Trait`
```

Comment:
@theemathas 

> Does this allow things like `impl<A, B, T: SomeTrait<A, B>> ReverseTrait<B, A> for T`?

That is allowed if `ReverseTrait` is defined in the current crate, but not otherwise.